src/cursor.rs

```diff
@@ -1,723 +1,720 @@
-/// positioned on the XNode and be ready to operate its target XEntry. If not, it means that the `CursorMut`
-/// is not able to touch the target `XEntry`. For this situation, the `CursorMut`
-/// can invoke `store` method which will expand the XArray to guarantee to reach the target position.
-///
-/// The `CursorMut` will also record all nodes passed from the head node to the target position
-/// in passed_node, thereby assisting it in performing some operations that involve searching upwards.
-///
-/// **Features for COW (Copy-On-Write).** The CursorMut guarantees that if it is exclusive,
-/// all nodes it traverses during the process are exclusively owned by the current XArray.
-/// If it finds that the node it is about to access is shared with another XArray due to a COW clone,
-/// it will trigger a COW to copy and create an exclusive node for access. Additionally,
-/// since it holds a mutable reference to the current XArray, it will not conflict with
-/// any other cursors on the XArray. CursorMut is set to exclusive when a modification
-/// is about to be performed
-///
-/// When a CursorMut doing write operation on XArray, it should not be affected by other CursorMuts
-/// or affect other Cursors.
-pub struct CursorMut<'a, I, L, M>
+/// **Features for COW (Copy-On-Write).** Due to COW, multiple `XArray`s can share the same piece
+/// of data. As a result, `CursorMut` does not always have exclusive access to the items stored in
+/// the `XArray`. However, just before performing the modification, `CursorMut` will create
+/// exclusive copies by cloning shared items, which guarantees the isolation of data stored in
+/// different `XArray`s.
+pub struct CursorMut<'a, I, M>
 where
     I: ItemEntry,
-    L: XLock,
     M: Into<XMark>,
 {
-    /// The `XArray` the cursor located in.
-    xa: &'a mut XArray<I, L, M>,
-    /// The target index of the cursor in the belonged `XArray`.
+    /// The `XArray` where the cursor locates.
+    xa: DormantMutRef<'a, XArray<I, M>>,
+    /// The target index of the cursor.
     index: u64,
-    /// Represents the current state of the cursor.
-    state: CursorState<'a, I, L, ReadWrite>,
-    /// Record add nodes passed from the head node to the target position.
-    /// The index is the height of the recorded node.
-    ancestors: SmallVec<[&'a XNode<I, L, ReadWrite>; MAX_HEIGHT]>,
-
-    is_exclusive: bool,
-
-    _marker: PhantomData<I>,
+    /// The state of the cursor.
+    state: CursorState<'a, I, ReadWrite>,
+    /// Ancestors of the leaf node (exclusive), starting from the root node and going down, until
+    /// the first node which is shared in multiple `XArray`s.
+    mut_ancestors: SmallVec<[NodeMutRef<'a, I>; MAX_HEIGHT]>,
+    /// Ancestors of the leaf node (exclusive), but only containing the nodes which are shared in
+    /// multiple `XArray`s, from the first one and going down.
+    ancestors: SmallVec<[&'a XNode<I>; MAX_HEIGHT]>,
```

Comment:
I'm not sure if the `ancestors` is necessary, it is a little bit weired to own two ancestor vector here. It seems to be used only to distinguish exclusive node and shared node. I think it can use a bool value like `is_exclusive` instead.

Reply:
Is it possible to add the method like `awaken_immutable` to obtain a &XNode? Because the only mutable reference is held in `NodeMutRef`, I think this method is safe. Then we can only store a vector of `NodeMutRef` and use `is_exclusive` to determine the state.

Reply:
We should not create `NodeMutRef`/`DormantMutRef` from an immutable reference. Currently, we can only create a `DormantMutRef` from an exclusive reference, as shown in the following code snippet:
```rust
    /// Creates a dormant mutable reference and returns both the original reference and the dormant
    /// reference, so that the original reference can continue to be used.
    pub fn new(val: &'a mut T) -> (&'a mut T, DormantMutRef<'a, T>) {
        let mut ptr = NonNull::from(val);
        // SAFETY: The original reference is still exclusive and can continue to be used.
        let new_val = unsafe { ptr.as_mut() };
        (
            new_val,
            DormantMutRef {
                ptr,
                _marker: PhantomData,
            },
        )
    }
```

If we allow it, we have to add another safety requirement for its `awaken` method:
```patch
diff --git a/src/borrow.rs b/src/borrow.rs
index f701f2a..699db54 100644
--- a/src/borrow.rs
+++ b/src/borrow.rs
@@ -33,6 +33,9 @@ impl<'a, T> DormantMutRef<'a, T> {
     ///
     /// The caller must ensure that the original reference and all its derived references
     /// (including derived dormant references) are now dead.
+    ///
+    /// The caller must also ensure the dormant references are created from exclusive references,
+    /// instead of shared ones.
     pub unsafe fn awaken(mut self) -> &'a mut T {
         // SAFETY: The safety requirements of the method ensure that the reference is valid and
         // exclusive.

```

Personally, I do not want to do this, because mixing exclusive and shared references does not seem right to me, and it introduces a lot of unnecessary risks. In principle, I think we should use safe Rust as much as possible, unless safe Rust is not powerful enough (which is not the case), or safe Rust is not efficient enough (there is a tradeoff, but in this case moving the logic to unsafe Rust does not bring any observable performance improvements either).

Reply:
I mean adding an `awaken_immutable()` method in DormantMutRef and NodeMutRef,  they are still created from mutable reference, but they can be awaken as an immutable reference. I think this method can be even a safe method because it seems allow derived reference exist at the same time. This does sound a bit odd so I am not sure is it feasible.

Reply:
You are right, this method is not elegant. The reason to introduce two vector is that when `CursorMut` traveling, it may go through either shared nodes or exclusive nodes. I originally have another idea that storing `&XEntry` in the ancestors, yet it will introduce overhead of transfering &XEntry to &mut XNode during usage, and it is also not compatible for sematics. Maybe currently use two vector is more reasonable.